### DISCUSSION TITLE
Column Seperator in Rofi Script

### DISCUSSION THREAD

**Discussion by blueray453** [davatorium/rofi#1871] [Q&A] [answered]
I have written a script to view window with workspace names.

The script looks like:

```
#!/usr/bin/env bash

json=$(dbus-send --print-reply=literal --session --dest=org.gnome.Shell /org/gnome/Shell/Extensions/GnomeUtilsWorkspaces org.gnome.Shell.Extensions.GnomeUtilsWorkspaces.GetWorkspaces)


current_workspace=$(echo $json | jq -r .current_workspace)
number_of_workspaces=$(echo $json | jq .number_of_workspaces)
all_normal_windows_of_workspaces=$(echo $json | jq .all_normal_windows_of_workspaces)
all_normal_windows_of_current_workspaces=$(echo $json | jq ".all_normal_windows_of_workspaces .$current_workspace" | jq -c '.[]' )

# icon|workspace_name|window_title

for item in ${all_normal_windows_of_current_workspaces}; do
  title=$(dbus-send --print-reply=literal --session --dest=org.gnome.Shell /org/gnome/Shell/Extensions/GnomeUtilsWindows org.gnome.Shell.Extensions.GnomeUtilsWindows.GetTitle uint32:$item)


  icon=$(dbus-send --print-reply=literal --session --dest=org.gnome.Shell /org/gnome/Shell/Extensions/GnomeUtilsWindows org.gnome.Shell.Extensions.GnomeUtilsWindows.GetIconFromWinID uint32:$item)

  # removing | from $title
  echo -en "${current_workspace}|${title//|/}|\0icon\x1ffolder\x1finfo\x1ftest\n"
done
```
For the dbus commands it need the extension https://github.com/blueray453/gnome-utils-by-blueray453

Anyway there are implementation details.

The issue is, I am calling it using:

    rofi -show gn -modi "gn:./test"

It looks like:

![Screenshot from 2023-07-21 20-12-43](https://github.com/davatorium/rofi/assets/855297/b19c4ae2-c1e8-4fcb-9c04-52c87dbb6e2e)

The issue is with the column separator.

When I asked last time you said:

> Rofi has no real 'columns ' support, it depends on (monospace) font and/or using spaces/tabs.

I am using monospace font.  **How can I use spaces/tabs to make column here?**

You also gave me link to https://davatorium.github.io/rofi/1.7.5/rofi-script.5/

The page does not describe how `aap\0icon\x1ffolder\x1finfo\x1ftest\n` works really. I mean, what is `\0`, `\x1f`. Where can i find details about these. I am assuming the column separator will also look like one of these.

Is there any way i can use `|` as column separator?

If you look at `echo -en "${current_workspace}|${title//|/}|\0icon\x1ffolder\x1finfo\x1ftest\n"`,  i have put the icons in last column but it is showing in the first column. How does it work actually?

**Comment by blueray453** [comment]
For example,

```
echo -en "Dot Nothing   This is title 1     \0icon\x1ffolder\x1finfo\x1ftest\n"
echo -en "Dot   This is title foxrot 1      \0icon\x1ffolder\x1finfo\x1ftest\n"
```
Here i put two tabs in each line. But it is not separating in column.

One more thing:

You can see that the icon is in `$icon` variable.

But `echo -en "app\0icon\x1f${icon}\x1finfo\x1ftest\n"` does not get the icon.

**Comment by DaveDavenport** [comment, marked as answer]
> The issue is with the column separator.

There  is no column separator.

>  Is there any way i can use | as column separator?

Asking the same question again does not change the answer:

> Rofi has no real 'columns ' support, it depends on (monospace) font and/or using spaces/tabs.


Using tabs and tab-stops to align text is a pretty basic thing (from typewriter era) that most software implement  (word, google docs, etc). 

```bash
echo -en "e\tbbub\tc\naap\tb\td" | rofi -dmenu
```
![image](https://github.com/davatorium/rofi/assets/406245/c019cfe7-33d6-4504-8163-48141d193c5c)

```bash
echo -en "e\tbbub\tc\naap\tb\td" | rofi -dmenu -theme-str 'element-text { tab-stops: [ 10em, 20em, 30em]; }'
```
![image](https://github.com/davatorium/rofi/assets/406245/d619c49e-2beb-469f-b4c6-26d073bd2184)

See documentation.


> The page does not describe how aap\0icon\x1ffolder\x1finfo\x1ftest\n works really. I mean, what is \0, \x1f. Where can i find details about these. 

They are characters, see man ascii.

`\0`  -> end of string
`\x1f` ->  unit separator
`\n` -> newline
`\t` -> tab

or see `rofi-script` manpage you linked:

> NULL character (\0) followed by a key, separator (\x1f) and value.

> I am assuming the column separator will also look like one of these.

Again the answer has not changed:

> Rofi has no real 'columns ' support, it depends on (monospace) font and/or using spaces/tabs.

There is no column separator

> If you look at echo -en "${current_workspace}|${title//|/}|\0icon\x1ffolder\x1finfo\x1ftest\n", i have put the icons in last column but it is showing in the first column. How does it work actually?


Because the icon widget is separated from the text and its location is specified by the theme (see man rofi-theme), and again:

> Rofi has no real 'columns ' support, it depends on (monospace) font and/or using spaces/tabs.


* https://en.wikipedia.org/wiki/Tab_stop

**Comment by blueray453** [comment]
I think i can handle it from here. The only issue remaining is:

```
echo -en "${current_workspace}\t${title}\0icon\x1ffolder\x1finfo\x1ftest\n"
echo -en "${current_workspace}\t${title}\0icon\x1fio.github.cboxdoerfer.FSearch\x1finfo\x1ftest\n"
```

Here if i replace `folder` or `io.github.cboxdoerfer.FSearch` with `${icon}` then it no longer works. do you have any idea why this might be happening?

**Reply by DaveDavenport** [reply]
Hard to say, if `${icon}` contains the right icon name, it should work. However with the information you gave me, I cannot even make a guess.

**Comment by blueray453** [comment]
If I run `echo -en "aap\0icon\x1f${icon}\n"` like:

```
#!/usr/bin/env bash

array="Alacritty vscodium firefox-esr io.github.cboxdoerfer.FSearch folder"

for icon in ${array}; do
    echo -en "aap\0icon\x1f${icon}\n"
done
```

The output of the command is:
```
$ ./test
firefox-esr
Alacritty
firefox-esr
vscodium
```

However, If I run `echo -en "aap\0icon\x1f${icon}\n"` like:

```
#!/usr/bin/env bash

json=$(dbus-send --print-reply=literal --session --dest=org.gnome.Shell /org/gnome/Shell/Extensions/GnomeUtilsWorkspaces org.gnome.Shell.Extensions.GnomeUtilsWorkspaces.GetWorkspaces)

workspace_names=$(echo $json | jq -c -r .workspace_names[])

for w in ${workspace_names}; do

all_normal_windows_of_nth_workspaces=$(echo $json | jq ".all_normal_windows_of_workspaces .$w" | jq -c '.[]' )

    for item in ${all_normal_windows_of_nth_workspaces}; do
        icon=$(dbus-send --print-reply=literal --session --dest=org.gnome.Shell /org/gnome/Shell/Extensions/GnomeUtilsWindows org.gnome.Shell.Extensions.GnomeUtilsWindows.GetIconFromWinID uint32:$item)
        echo -en "aap\0icon\x1f${icon}\n"
    done
done
```

The output is like:

```
./test
appicon   firefox-esr
appicon   Alacritty
appicon   firefox-esr
appicon   vscodium
```
Not sure what to do here.

**Reply by DaveDavenport** [reply]
the output of your first script is:
![image](https://github.com/davatorium/rofi/assets/406245/366a6f62-0177-4a27-bfef-5bc7db8be9bc)

if I pass it to rofi, I see icons.(atleast the ones I have).

can you run the second script with:

```bash
./script.sh > mylog.txt
```
and upload the mylog file, without changing it. (if nothing private is in there).

**Reply by blueray453** [reply]
Here is the log file.

[mylog.txt](https://github.com/davatorium/rofi/files/12136575/mylog.txt)

**Reply by DaveDavenport** [reply]
If I try that file, I see that rofi tries to resolve the icons:
```
(process:18267): Helpers.IconFetcher-DEBUG: 15:35:48.216: Query:    firefox-esr(50)
(process:18267): Helpers.IconFetcher-DEBUG: 15:35:48.216: starting up icon fetching thread.
(process:18267): Helpers.IconFetcher-DEBUG: 15:35:48.216: Query:    Alacritty(50)
(process:18267): Helpers.IconFetcher-DEBUG: 15:35:48.217: starting up icon fetching thread.
(process:18267): Helpers.IconFetcher-DEBUG: 15:35:48.217: Query:    Alacritty(50)
(process:18267): Helpers.IconFetcher-DEBUG: 15:35:48.217: Query:    Alacritty(50)
(process:18267): Helpers.IconFetcher-DEBUG: 15:35:48.217: Query:    firefox-esr(50)
(process:18267): Helpers.IconFetcher-DEBUG: 15:35:48.217: Query:    firefox-esr(50)
(process:18267): Helpers.IconFetcher-DEBUG: 15:35:48.218: Query:    vscodium(50)
(process:18267): Helpers.IconFetcher-DEBUG: 15:35:48.218: starting up icon fetching thread.
(process:18267): Helpers.IconFetcher-DEBUG: 15:35:48.232: failed to get icon    firefox-esr(50x50): n/a
(process:18267): Helpers.IconFetcher-DEBUG: 15:35:48.232: failed to get icon    Alacritty(50x50): n/a
(process:18267): Helpers.IconFetcher-DEBUG: 15:35:48.233: failed to get icon    vscodium(50x50): n/a
```

I think the icon names are not right.
Alacritty should be 'com.alacritty.Alacritty'  if I fix this in mylog.txt it works.

**Reply by blueray453** [reply]
Here is a screenshot running the code in my machine:

![new_screen](https://github.com/davatorium/rofi/assets/855297/a2a79f75-63de-4638-afda-ce34c7e51b69)

Please let me give you the `echo -en "aap\0icon\x1f${icon}\n" | head -1 | od -c` output.

in case of:

```
#!/usr/bin/env bash

array="Alacritty vscodium firefox-esr io.github.cboxdoerfer.FSearch folder"

[for icon in ${array}; do
    echo -en "aap\0icon\x1f${icon}\n"
    echo -en "aap\0icon\x1f${icon}\n" | head -1 | od -c
done](url)
```

The output is:

```
$ ./test
aapiconAlacritty
0000000   a   a   p  \0   i   c   o   n 037   A   l   a   c   r   i   t
0000020   t   y  \n
0000023
aapiconvscodium
0000000   a   a   p  \0   i   c   o   n 037   v   s   c   o   d   i   u
0000020   m  \n
0000022
aapiconfirefox-esr
0000000   a   a   p  \0   i   c   o   n 037   f   i   r   e   f   o   x
0000020   -   e   s   r  \n
0000025
aapiconio.github.cboxdoerfer.FSearch
0000000   a   a   p  \0   i   c   o   n 037   i   o   .   g   i   t   h
0000020   u   b   .   c   b   o   x   d   o   e   r   f   e   r   .   F
0000040   S   e   a   r   c   h  \n
0000047
aapiconfolder
0000000   a   a   p  \0   i   c   o   n 037   f   o   l   d   e   r  \n
0000020
```
In case of 

```
#!/usr/bin/env bash

json=$(dbus-send --print-reply=literal --session --dest=org.gnome.Shell /org/gnome/Shell/Extensions/GnomeUtilsWorkspaces org.gnome.Shell.Extensions.GnomeUtilsWorkspaces.GetWorkspaces)

workspace_names=$(echo $json | jq -c -r .workspace_names[])

for w in ${workspace_names}; do

all_normal_windows_of_nth_workspaces=$(echo $json | jq ".all_normal_windows_of_workspaces .$w" | jq -c '.[]' )

    for item in ${all_normal_windows_of_nth_workspaces}; do
        icon=$(dbus-send --print-reply=literal --session --dest=org.gnome.Shell /org/gnome/Shell/Extensions/GnomeUtilsWindows org.gnome.Shell.Extensions.GnomeUtilsWindows.GetIconFromWinID uint32:$item)
        echo -en "aap\0icon\x1f${icon}\n"
        echo -en "aap\0icon\x1f${icon}\n" | head -1 | od -c
    done
done
```
The output is:

```
./test
aapicon   firefox-esr
0000000   a   a   p  \0   i   c   o   n 037               f   i   r   e
0000020   f   o   x   -   e   s   r  \n
0000030
aapicon   Alacritty
0000000   a   a   p  \0   i   c   o   n 037               A   l   a   c
0000020   r   i   t   t   y  \n
0000026
aapicon   firefox-esr
0000000   a   a   p  \0   i   c   o   n 037               f   i   r   e
0000020   f   o   x   -   e   s   r  \n
0000030
aapicon   firefox-esr
0000000   a   a   p  \0   i   c   o   n 037               f   i   r   e
0000020   f   o   x   -   e   s   r  \n
0000030
aapicon   vscodium
0000000   a   a   p  \0   i   c   o   n 037               v   s   c   o
0000020   d   i   u   m  \n
0000025
```

**Reply by DaveDavenport** [reply]
So dbus includes some extra spaces that need to be trimmed off?
(it seems to prepend 3 spaces)

**Reply by blueray453** [reply]
I think you are right. Do you have any idea how to remove the white space?

**Reply by DaveDavenport** [reply]
https://www.cyberciti.biz/faq/bash-remove-whitespace-from-string/?

```bash
# 1. Turn it on
shopt -s extglob
 
# 2. Set variable and delete leading whiltespaces 
output="    This is a test"
output="${output##*( )}"
echo "=${output}="
 
# 3. Turn it off
shopt -u extglob
```
I think the `${icon##*( )}` might work?

**Reply by blueray453** [reply]
icon=$(dbus-send --print-reply=literal --session --dest=org.gnome.Shell /org/gnome/Shell/Extensions/GnomeUtilsWindows org.gnome.Shell.Extensions.GnomeUtilsWindows.GetIconFromWinID uint32:$item | xargs)

also solve the problem.
